### PR TITLE
feat: Rect and DirectStyleRule components

### PR DESCRIPTION
**Pull request by andrew-codes**
Rect retrieves observable dimensions of a child element. Direct style rule enables applying a style rule directly to a child. Both have e2e tests.

### CODE REVIEW

File: packages/components/src/Rect/README.mdx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    )}
+
+  </Rect>
+</Playground>
+
+## Component API
+
+<PropsTable of={Rect} />
+
+## Theme API
```
Comment (TheKeithStewart) [inline packages/components/src/Rect/README.mdx]: Does this need a Theme API? Doesn't seem to apply here.

File: packages/components/src/Rect/README.mdx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+          >
+            {JSON.stringify(rect, null, 2)}
+        </div>
+      </div>
+    )}
+
+  </Rect>
+</Playground>
+
+## Component API
```
Comment (TheKeithStewart) [inline packages/components/src/Rect/README.mdx]: The API table doesn't seem to be getting generated ([docs](https://deploy-preview-7--inspiring-varahamihira-92fa33.netlify.com/components/Rect#component-api))

File: packages/components/src/DirectStyleRule/README.mdx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+route: /components/DirectStyleRule
+---
+
+import { Fragment } from 'react';
+import { Playground, PropsTable, ThemeDefinitionTable } from '@versionone/docz';
+import { DirectStyleRule } from './';
+
+## Usage
+
+Apply a style rule directly to children without rendering an intermediate element.
```
Comment (walkerrandolphsmith) [inline packages/components/src/DirectStyleRule/README.mdx]: We should describe the behavior when you apply styles that have already been defined on the instance. Which value wins the DirectStyleRule or the thing it wraps? Could we get into a situation similar to css rules overriding one another?

File: packages/components/src/DirectStyleRule/DirectStyleRule.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import { StyleProvider } from '../StyleProvider';
+
+const DirectStyleRule = ({
+  children,
+  styleRule,
+  passThroughProps,
+  ...rest
+}) => {
+  const Impl = createComponent(styleRule, 'span', passThroughProps);
+  const instance = isString(children) ? false : React.Children.only(children);
```
[the remark this answers is not in the document]
Comment (andrew-codes) [reply packages/components/src/DirectStyleRule/DirectStyleRule.js]: I believe that `React.Children.only()` will throw an error if there is more than one child, so the propType check (which wouldn't be in production) is not the only fail-safe.

In terms of should we handle multiple children, I do not think it makes sense to do so; unless the style rule is applied directly to all children. I guess that can make sense. I'm curious to hear more thoughts on this.

File: packages/components/src/DirectStyleRule/README.mdx (hunk 3 above)
Comment (andrew-codes) [reply packages/components/src/DirectStyleRule/README.mdx]: I **think** we can get into an overriding issue between the direct styles and the component's styles. This is not ideal. I feel the component's styles should always win in these cases; not the Directly applied style rule. Shedding some light on the implementation may help root out an answer:

DirectStyleRule effectively appends the generated CSS class names its rule creates to the underlying element. With this said, none of our components currently accept className. Assuming we keep this direction, which I feel we should, then `DirectStyleRule` is only relevant to components created via `createComponent` AND with className as a passThroughProp. Hopefully, that helps answer your question. We can most certainly iterate on `DirectStyleRule`.

Comment (walkerrandolphsmith) [reply packages/components/src/DirectStyleRule/README.mdx]: Honestly, I'd like to see an explicit need for this component before it exists. In the case of say AvatarGroup we might not need it today for a feature, but we know we will need it for say a future set of work. In this case I can only speculate I might need it... I'd say until there is a strong case of needing it (more than just a hypothetical case) I wouldn't advocate using it.

**Pull request closed without merging**